sdk/include/compartment-macros.h

```diff
@@ -0,0 +1,174 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+#include <cdefs.h>
+
+/**
+ * Provide a capability of the type `volatile type *` referring to the MMIO
+ * region exported in the linker script with `name` as its name.  This macro
+ * can be used only in code (it cannot be used to initialise a global).
+ */
+#define MMIO_CAPABILITY(type, name)                                            \
+	({                                                                         \
+		volatile type *ret; /* NOLINT(bugprone-macro-parentheses) */           \
+		__asm(".ifndef __import_mem_" #name "\n"                               \
+		      "  .type     __import_mem_" #name ",@object\n"                   \
+		      "  .section  .compartment_imports." #name                        \
+		      ",\"awG\",@progbits," #name ",comdat\n"                          \
+		      "  .globl    __import_mem_" #name "\n"                           \
+		      "  .p2align  3\n"                                                \
+		      "__import_mem_" #name ":\n"                                      \
+		      "  .word __export_mem_" #name "\n"                               \
+		      "  .word __export_mem_" #name "_end - __export_mem_" #name "\n"  \
+		      "  .size __import_mem_" #name ", 8\n"                            \
+		      " .previous\n"                                                   \
+		      ".endif\n"                                                       \
+		      "1:"                                                             \
+		      "  auipcc  %0,"                                                  \
+		      "      %%cheri_compartment_pccrel_hi(__import_mem_" #name ")\n"  \
+		      "  clc     %0, %%cheri_compartment_pccrel_lo(1b)(%0)\n"          \
+		      : "=C"(ret));                                                    \
+		ret;                                                                   \
+	})
+
+/**
+ * Macro to test whether a device with a specific name exists in the board
+ * definition for the current target.
```

Comment:
I know it's not part of this PR, but this comment really should be expanded to include something about this macro being useful *only* within a preprocessor expression.  (It's probably best to limit it to preprocessor evaluation contexts rather than pursue terrors like http://hmijailblog.blogspot.com/2016/03/an-isdefined-c-macro-to-check-whether.html ...)